File: solutions/0907. Sum of Subarray Minimums/0907.java
class Solution {
  public int sumSubarrayMins(int[] A) {
    final int n = A.length;
    final int kMod = (int) 1e9 + 7;

    long ans = 0;
    int[] prev = new int[n];
    int[] next = new int[n];
    Stack<Integer> stack = new Stack<>();

    Arrays.fill(prev, -1);
    Arrays.fill(next, n);

    for (int i = 0; i < A.length; ++i) {
      while (!stack.isEmpty() && A[stack.peek()] > A[i]) {
        final int index = stack.pop();
        next[index] = i;
      }
      if (!stack.isEmpty())
        prev[i] = stack.peek();
      stack.add(i);
    }

    for (int i = 0; i < A.length; ++i)
      ans = (ans + (long) A[i] * (i - prev[i]) * (next[i] - i)) % kMod;

    return (int) ans;
  }
}

File: solutions/0907. Sum of Subarray Minimums/0907.py
class Solution:
  def sumSubarrayMins(self, A: List[int]) -> int:
    n = len(A)
    kMod = int(1e9 + 7)

    ans = 0
    prev = [-1] * n
    next = [n] * n
    stack = []

    for i, a in enumerate(A):
      while stack and A[stack[-1]] > a:
        index = stack.pop()
        next[index] = i
      if stack:
        prev[i] = stack[-1]
      stack.append(i)

    for i, a in enumerate(A):
      ans = (ans + a * (i - prev[i]) * (next[i] - i)) % kMod

    return ans
